From 8bebe4d88911d369184174802e4176beffe71fdd Mon Sep 17 00:00:00 2001
From: Alexander Smirnov <asmirnov@embeddedalley.com>
Date: Mon, 29 Dec 2008 21:24:50 +0300
Subject: [PATCH] Add mx27ads board specific code for keypad

Add mx27ads board specific code for keypad configurable via
CONFIG_KEYBOARD_MXC.

Signed-off-by: Alexander Smirnov <asmirnov@embeddedalley.com>
---
 arch/arm/mach-mx2/mx27ads.c |   54 +++++++++++++++++++++++++++++++++++++++++++
 1 files changed, 54 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-mx2/mx27ads.c b/arch/arm/mach-mx2/mx27ads.c
index 5adc338..8d3fd1b 100644
--- a/arch/arm/mach-mx2/mx27ads.c
+++ b/arch/arm/mach-mx2/mx27ads.c
@@ -30,6 +30,7 @@
 #include <linux/spi/spi.h>
 #include <linux/dma-mapping.h>
 #include <linux/delay.h>
+#include <linux/input.h>
 
 #include <asm/mach-types.h>
 #include <asm/mach/arch.h>
@@ -53,6 +54,7 @@
 #include <mach/mxc_ehci.h>
 #include <mach/ulpi.h>
 #include <mach/arc_otg.h>
+#include <mach/keypad.h>
 
 #include "devices.h"
 
@@ -118,6 +120,55 @@ static struct platform_device mx27ads_nor_mtd_device = {
 	.resource = &mxc_flash_resource,
 };
 
+#if defined(CONFIG_KEYBOARD_MXC)
+
+static u16 mxckpd_keycodes[] = {
+	KEY_KP9, KEY_LEFTSHIFT, KEY_0, KEY_KPASTERISK, KEY_RECORD, KEY_POWER,
+	KEY_KP8, KEY_9, KEY_8, KEY_7, KEY_KP5, KEY_VOLUMEDOWN,
+	KEY_KP7, KEY_6, KEY_5, KEY_4, KEY_KP4, KEY_VOLUMEUP,
+	KEY_KP6, KEY_3, KEY_2, KEY_1, KEY_KP3, KEY_DOWN,
+	KEY_BACK, KEY_RIGHT, KEY_ENTER, KEY_LEFT, KEY_HOME, KEY_KP2,
+	KEY_END, KEY_F2, KEY_UP, KEY_F1, KEY_F4, KEY_KP1,
+};
+
+static struct keypad_data evb_6_by_6_keypad = {
+	.rowmax = 6,
+	.colmax = 6,
+	.numkeys = ARRAY_SIZE(mxckpd_keycodes),
+	.delay = 2,
+	.matrix = mxckpd_keycodes,
+};
+
+static struct resource mxc_kpp_resources[] = {
+	[0] = {
+		.start = MXC_INT_KPP,
+		.end = MXC_INT_KPP,
+		.flags = IORESOURCE_IRQ,
+	}
+};
+
+static struct platform_device mx27ads_keypad_device = {
+	.name = "mxc_keypad",
+	.id = 0,
+	.num_resources = ARRAY_SIZE(mxc_kpp_resources),
+	.resource = mxc_kpp_resources,
+	.dev = {
+		.platform_data = &evb_6_by_6_keypad,
+	},
+};
+
+void gpio_keypad_active(void)
+{
+}
+EXPORT_SYMBOL(gpio_keypad_active);
+
+void gpio_keypad_inactive(void)
+{
+}
+EXPORT_SYMBOL(gpio_keypad_inactive);
+
+#endif /* CONFIG_MXC_KEYBOARD */
+
 #ifdef CONFIG_MTD_NAND
 static struct mxc_nand_platform_data mx27ads_nand_board_info = {
 	.width = 1,
@@ -887,6 +938,9 @@ static struct platform_device *platform_devices[] __initdata = {
 #ifdef CONFIG_SND
 	&mx27ads_alsa_sound_device,
 #endif
+#ifdef CONFIG_KEYBOARD_MXC
+	&mx27ads_keypad_device,
+#endif /* KEYBOARD_MXC */
 };
 
 static void __init mx27ads_board_init(void)
-- 
1.5.5.1

